fs: fix cp, cpSync handle existing symlinks

jakecastelli opened the pull request:
Fixes: #58468 

<!--
Before submitting a pull request, please read:

- the CONTRIBUTING guide at https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md
- the commit message formatting guidelines at
  https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

If you believe this PR should be highlighted in the Node.js CHANGELOG
please add the `notable-change` label.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

dario-piotrowicz commented on lib/internal/fs/cp/cp-sync.js:
```diff
@@ -199,7 +199,7 @@ function onLink(destStat, src, dest, verbatimSymlinks) {
     resolvedDest = resolve(dirname(dest), resolvedDest);
   }
-  if (isSrcSubdir(resolvedSrc, resolvedDest)) {
+  if (fsBinding.internalModuleStat(src) === 1 && isSrcSubdir(resolvedSrc, resolvedDest)) {
```
could we maybe use a temporary variable to make the code more understandable? :slightly_smiling_face: 
```suggestion
  const srcIsDir = fsBinding.internalModuleStat(src) === 1;
  if (srcIsDir && isSrcSubdir(resolvedSrc, resolvedDest)) {
```

dario-piotrowicz commented:
PS: I don't think that this PR fully addressed https://github.com/nodejs/node/issues/58468 since the issue mentions the bug both in `cp` and `cpSync`, @jakecastelli could you solve the bug in both functions? :slightly_smiling_face:  (or alternatively I can split the issue into two if you prefer?)

jakecastelli commented:
Thanks for the pointer, I was a bit running late last night and didn't pay close attention to the `cp` repo code, I have fixed both places with newly added tests, and reverted https://github.com/nodejs/node/commit/52430b9839c31f846a788125efda82887d115e21 🙏 PTAL

dario-piotrowicz commented on test/parallel/test-fs-cp.mjs:
```diff
@@ -248,6 +248,33 @@ function nextdir(dirname) {
+  cpSync(src, dest, mustNotMutateObjectDeep({ recursive: true }));
+}
+
+{
```
Could you add a comment for this one too? :slightly_smiling_face: :pray:

dario-piotrowicz commented on test/parallel/test-fs-cp.mjs:
```diff
@@ -248,6 +248,33 @@ function nextdir(dirname) {
   );
 }
 
+// It allows copying when is not a directory
```
mh.... I am really not sure about this comment, could you clarify it? :pray: 

maybe something like
```suggestion
// It allows copying symlinks in src to locations in dest with existing synlinks not pointing to a directory.
```
?

jakecastelli replied:
sgtm!

dario-piotrowicz reviewed (approved):
Looks great! thanks a lot again @jakecastelli :pray: 

I just left a couple of comments about the tests comments, I hope those can be addressed before landing this :slightly_smiling_face::pray:

dario-piotrowicz commented on test/parallel/test-fs-cp.mjs:
```diff
@@ -248,6 +248,33 @@ function nextdir(dirname) {
   );
 }
 
+// It allows copying symlinks in src to locations in dest with existing synlinks not pointing to a directory.
```
Sure :slightly_smiling_face:, looks good to me, but could you add a comment to the next test block too? :sweat_smile::pray:

dario-piotrowicz replied:
yes, ideally you can specify the sync/callback variant in the comment too :slightly_smiling_face: 

but regardless of that, even if they were 100% identical comments I would still find that valuable, since every block should have its own comment IMHO

Basically when I check the test blocks I read the comment right above each block to get context on what it is testing, so in this case if I were to see these two blocks as they are now I would assume that the developer just forgot to put a comment on the second block (I would not think that the comment above the first block refers to both blocks), is that just me? :sweat_smile: :thinking:

dario-piotrowicz replied:
Basically the way I see it is that each test block is like a test runner `test()`/`it()` and each of those do have their own text description :slightly_smiling_face: 

PS: for reasons such as this I think it would be clearer and nicer to use the test runner across the board for all our tests... but from experience I know many disagree with that opinion :face_holding_back_tears:

jakecastelli replied:
Sounds good to me 👍 I have specified the sync and cb version in each comment block

dario-piotrowicz commented:
@aduh95 do you think we could/should just revert the change only for v22?

dario-piotrowicz commented:
> @dario-piotrowicz it will be included only if we can build and the tests are passing. 

got it, thanks :slightly_smiling_face::+1: 

> If it's not worth backporting, we should add the [ dont-land-on-v22.x](https://github.com/nodejs/node/labels/dont-land-on-v22%2Ex) PRs that should not land on the v22.x-staging branch and should not be released in v22.x. instead

Sorry you're right, I was asking more in the sense of this being a small bug fix, so if it were problematic in v22 we could avoid the backport, but yeah ideally it should land in v22 as well...
